### PR TITLE
Allow `split_out` to be `None`, which then defaults to `1` in `groupby().aggregate()`

### LINKED ISSUE #9490: Can no longer provide `split_out=None` as a default in `groupby.aggregate`
**What happened**:

Since a combination of #9453 and #9302, one can no longer provide `split_out=None` to `groupby.aggregate` to obtain the default `split_out` value (eventually `split_out=None` defaults to `split_out=1` deep inside `apply_concat_apply`).

(cc @rjzamora, @ian-r-rose)

**What you expected to happen**:

`split_out=None` should default to something sensible. The question arises as to whether one should change the API of all the aggregate functions to default to `split_out=None` rather than `split_out=1` or if we should just handle it locally. (The lack of true option-types in Python bites again 😢)

**Minimal Complete Verifiable Example**:

```python
import dask.dataframe
import pandas as pd

df = pd.DataFrame({"grouping": [1, 2, 1, 1, 3, 2, 1, 1], "a": list(range(8))})

ddf = dask.dataframe.from_pandas(df)

groups = ddf.groupby("grouping").aggregate("sum", split_out=None).compute()

TypeError                                 Traceback (most recent call last)
Input In [5], in <cell line: 1>()
----> 1 ddf.groupby("grouping").aggregate("sum", split_out=None).compute()

File ~/Documents/src/dask/dask/dataframe/groupby.py:2265, in DataFrameGroupBy.aggregate(self, arg, split_every, split_out, shuffle)
   2262 if arg == "size":
   2263     return self.size()
-> 2265 return super().aggregate(
   2266     arg, split_every=split_every, split_out=split_out, shuffle=shuffle
   2267 )

File ~/Documents/src/dask/dask/dataframe/groupby.py:1685, in _GroupBy.aggregate(self, arg, split_every, split_out, shuffle)
   1682 @_aggregate_docstring()
   1683 def aggregate(self, arg, split_every=None, split_out=1, shuffle=None):
   1684     if shuffle is None:
-> 1685         if split_out > 1:
   1686             shuffle = shuffle or config.get("shuffle", None) or "tasks"
   1687         else:

TypeError: '>' not supported between instances of 'NoneType' and 'int'
```

A similar issue occurs in `apply_concat_apply` as well:

```
dask.dataframe.core.aca(ddf, chunk=len, aggregate=sum, combine=sum, sort=True)
in apply_concat_apply(args, chunk, aggregate, combine, meta, token, chunk_kwargs, aggregate_kwargs, combine_kwargs, split_every, split_out, split_out_setup, split_out_setup_kwargs, sort, ignore_index, **kwargs)
   6447 # Handle sort behavior
   6448 if sort is not None:
-> 6449     if sort and split_out > 1:
   6450         raise NotImplementedError(
   6451             "Cannot guarantee sorted keys for `split_out>1`."
   6452             " Try using split_out=1, or grouping with sort=False."
   6453         )
   6454     aggregate_kwargs = aggregate_kwargs or {}

TypeError: '>' not supported between instances of 'NoneType' and 'int'
```

### PR DESCRIPTION
**Pull request by ian-r-rose**
- [x] Closes #9490 
- [ ] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Review by jrbourbeau — COMMENTED** [review]
Thanks @ian-r-rose! Would you mind adding a small test (e.g. similar to the example in https://github.com/dask/dask/issues/9490)?

**Comment by jrbourbeau** [comment]
Also, should we deprecate allowing `None` here?

**Comment by ian-r-rose** [comment]
> Thanks @ian-r-rose! Would you mind adding a small test (e.g. similar to the example in #9490)?

I was a bit uncomfortable adding a test since I wasn't sure if we actually wanted to support `None` here. I'd be curious to hear from @wence- if there are any things that would become hugely onerous if we dropped that support.

**Comment by jrbourbeau** [comment]
Agreed, it'd be good to get thoughts from @wence- on his particular use case. That said, do we advertise anywhere that `split_out` can be `None`? From a brief skim of the API docs, it looks like we're consistent in saying it's an `int`

**Comment by ian-r-rose** [comment]
I only see `None` documented in one place (sort of), which is the docstring for `groupby.aggregate`: https://docs.dask.org/en/latest/generated/dask.dataframe.groupby.DataFrameGroupBy.aggregate.html#dask.dataframe.groupby.DataFrameGroupBy.aggregate

Even then, it's saying it's an optional arg, which is a bit ambiguous.

**Comment by ian-r-rose** [comment]
The implementation for `apply_concat_apply` clearly anticipates a `None` value, but I might consider that an internal API.

**Comment by wence-** [comment]
> The implementation for `apply_concat_apply` clearly anticipates a `None` value, but I might consider that an internal API.

That also has a bug :( (`sort=True, split_out=None` => `TypeError`)

**Comment by wence-** [comment]
> > Thanks @ian-r-rose! Would you mind adding a small test (e.g. similar to the example in #9490)?
> 
> I was a bit uncomfortable adding a test since I wasn't sure if we actually wanted to support `None` here. I'd be curious to hear from @wence- if there are any things that would become hugely onerous if we dropped that support.

I don't think it would be problematic. I noticed because in `dask_cudf` we explicitly test `groupby().agg` with `split_out=None`. But that is in tests so I can remove it.

The maintenance issue here is somehow that there are lots of public APIs that call into the few "semi-private" APIs. So there are lots of places for the defaults to go out of whack.

One way to partially solve this is to _not provide default values_ for arguments to internal APIs, and promote the default handling to the public callers. I generally think that internal APIs should not do argument defaulting, since all that business logic handling is not really germane to the usage of the function.

It does make calling these functions slightly more verbose (especially given that they all have many arguments), though one can utilise my favourite python argument list approach of required keyword-only args:
```
def aca(*, split_out, split_every, ...)
```

**Comment by ian-r-rose** [comment]
> though one can utilise my favourite python argument list approach of required keyword-only args:

If I ever have a week of nothing to do, I'd love to make 90% of the APIs here make use of keyword-only args.

**Comment by wence-** [comment]
> Also, should we deprecate allowing `None` here?

I am inclined to say yes. I will change things on the dask_cudf side.

**Comment by rjzamora** [comment]
My concern with dropping support for `split_out=None` is that we **may** find a reasonable way to automatically infer a good value for `split_out` in the future.  For example, perhaps the `Groupby` class itself should **optionally** infer group cardinality up-front.  In that case, we would probably not want to override `split_out` if it was already set to a specific integer value.

This hypothetical concern may not be realistic.

**Comment by ian-r-rose** [comment]
> My concern with dropping support for `split_out=None` is that we **may** find a reasonable way to automatically infer a good value for `split_out` in the future. For example, perhaps the `Groupby` class itself should **optionally** infer group cardinality up-front. In that case, we would probably not want to override `split_out` if it was already set to a specific integer value.
> 
> This hypothetical concern may not be realistic.

I think it's a realistic hypothetical, but I guess that in that case I would advocate for expanding the accepted types to include `None`, or a sentinel value like the pandas `NoDefault` when it actually comes up. That change shouldn't be API-breaking for, except to the extent that we might want to make inference default.

File: dask/dataframe/groupby.py [hunk 1]
```diff
@@ -1681,6 +1681,14 @@ def get_group(self, key):
 
     @_aggregate_docstring()
     def aggregate(self, arg, split_every=None, split_out=1, shuffle=None):
+        split_out = 1 if split_out is None else split_out
```
Comment (jrbourbeau) [inline dask/dataframe/groupby.py:1684]: Given the logic below, it looks like we want to drop this line 

```suggestion
```

File: dask/dataframe/groupby.py [hunk 2]
```diff
@@ -1681,6 +1681,14 @@ def get_group(self, key):
...
+        if split_out is None:
+            warnings.warn(
+                "split_out=None is deprecated, please use a positive integer, "
+                "or allow the default of 1",
+                category=DeprecationWarning,
```
Comment (jrbourbeau) [inline dask/dataframe/groupby.py:1689]: Maybe `FutureWarning` instead as this is intended for end-users? 

```suggestion
                category=FutureWarning,
```

Comment (jrbourbeau) [reply dask/dataframe/groupby.py:1689]: Also, could we add a small test to make sure this warning is being emitted as expected?

File: dask/dataframe/groupby.py (hunk 1 above)
Comment (ian-r-rose) [reply dask/dataframe/groupby.py:1684]: Whoops, getting careless

**Review by jrbourbeau — APPROVED** [review]
Thanks @ian-r-rose!
